### PR TITLE
Modified PathInternal.Unix for great perf increase

### LINKED ISSUE #3515: PathInternal.Unix.HasIllegalCharacters far slower than its Windows equivalent
almost certainly related to #3512

Path.Combine and Path.GetFileName are both far slower than their windows counterparts. 

Linux results:
```
<assemblies>
  <assembly name="System.Runtime.Extensions.Tests.dll" environment="64-bit .NET (unknown version) [collection-per-class, parallel (1 threads)]" test-framework="xUnit.net 2.1.0.3168" run-date="2015-09-29" run-time="06:42:26" total="6" passed="6" failed="0" skipped="0" time="74.619" errors="0">
    <errors />
    <collection total="6" passed="6" failed="0" skipped="0" name="Test collection for System.Runtime.Extensions.Tests.Perf_Path" time="71.430">
      <test name="System.Runtime.Extensions.Tests.Perf_Path.Combine(innerIterations: 10000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="Combine" time="5.9924355" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.Combine(innerIterations: 20000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="Combine" time="11.906245" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.Combine(innerIterations: 30000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="Combine" time="18.788925" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetFileName(innerIterations: 10000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetFileName" time="5.8670823" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetFileName(innerIterations: 20000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetFileName" time="11.5486548" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetFileName(innerIterations: 30000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetFileName" time="17.3264216" result="Pass" />
    </collection>
  </assembly>
</assemblies>
```

Windows Results:
```
<assemblies>
  <assembly name="System.Runtime.Extensions.Tests.dll" environment="64-bit .NET (unknown version) [collection-per-class, parallel (8 threads)]" test-framework="xUnit.net 2.1.0.3168" run-date="2015-09-28" run-time="16:00:05" total="3" passed="3" failed="0" skipped="0" time="3.491" errors="0">
    <errors />
    <collection total="3" passed="3" failed="0" skipped="0" name="Test collection for System.Runtime.Extensions.Tests.Perf_Path" time="3.313">
      <test name="System.Runtime.Extensions.Tests.Perf_Path.Combine(innerIterations: 10000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="Combine" time="1.060862" result="Pass">
      <test name="System.Runtime.Extensions.Tests.Perf_Path.Combine(innerIterations: 20000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="Combine" time="1.0923945" result="Pass">
      <test name="System.Runtime.Extensions.Tests.Perf_Path.Combine(innerIterations: 30000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="Combine" time="1.1601604" result="Pass">
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetFileName(innerIterations: 10000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetFileName" time="1.0321461" result="Pass">
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetFileName(innerIterations: 20000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetFileName" time="1.0844894" result="Pass">
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetFileName(innerIterations: 30000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetFileName" time="1.1291452" result="Pass">
    </collection>
  </assembly>
</assemblies>
```

I believe this is caused by the PathInternal.Unix.HasIllegalCharacters function that is called by nearly every Path function. The code for the method in its current state that gave the above timing information is this:
```
        internal static bool HasIllegalCharacters(string path, bool checkAdditional = false)
        {
            Debug.Assert(path != null);
            foreach (char c in path)
            {
                // Same as InvalidPathChars, unrolled here for performance
                if (c == InvalidPathChar)
                    return true;
            }

            return false;
        }
```

By changing the function to be more similar to its Windows equivalent as follows:
```
        internal static bool HasIllegalCharacters(string path, bool checkAdditional = false)
        {
            Debug.Assert(path != null);
            return path.IndexOf('\0') >= 0;
        }
```

The perf results are significantly faster:
```
<assemblies>
  <assembly name="System.Runtime.Extensions.Tests.dll" environment="64-bit .NET (unknown version) [collection-per-class, parallel (1 threads)]" test-framework="xUnit.net 2.1.0.3168" run-date="2015-09-29" run-time="06:56:16" total="6" passed="6" failed="0" skipped="0" time="15.210" errors="0">
    <errors />
    <collection total="6" passed="6" failed="0" skipped="0" name="Test collection for System.Runtime.Extensions.Tests.Perf_Path" time="11.783">
      <test name="System.Runtime.Extensions.Tests.Perf_Path.Combine(innerIterations: 10000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="Combine" time="1.8116721" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.Combine(innerIterations: 20000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="Combine" time="2.4075532" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.Combine(innerIterations: 30000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="Combine" time="2.2146143" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetFileName(innerIterations: 10000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetFileName" time="1.3606556" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetFileName(innerIterations: 20000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetFileName" time="1.595163" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetFileName(innerIterations: 30000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetFileName" time="2.3932058" result="Pass" />
    </collection>
  </assembly>
</assemblies>
```

### PR DESCRIPTION
**Pull request by ianhays**
resolves #3515, #3515

### CODE REVIEW

File: src/Common/src/System/IO/PathInternal.Unix.cs [hunk 1]
```diff
@@ -16,15 +16,7 @@ internal static partial class PathInternal
...
-
-            foreach (char c in path)
-            {
-                // Same as InvalidPathChars, unrolled here for performance
-                if (c == InvalidPathChar)
-                    return true;
-            }
-
-            return false;
+            return path.IndexOf('\0') >= 0;
```
Comment (stephentoub) [inline src/Common/src/System/IO/PathInternal.Unix.cs]: InvalidPathChar instead of '\0' ?

Comment (ianhays) [inline src/Common/src/System/IO/PathInternal.Unix.cs]: It seemed like an unnecessary redirection so I hard-coded the value instead, but I can change it back to "InvalidPathChar" if that would be preferred.

Will the compiler inline the value?

Comment (stephentoub) [inline src/Common/src/System/IO/PathInternal.Unix.cs]: It's a const and will result in exactly the same IL.

Comment (ianhays) [inline src/Common/src/System/IO/PathInternal.Unix.cs]: fixed

**Comment by stephentoub** [comment]
LGTM

**Pull request merged**